sys/shell/shell.c

```diff
@@ -58,164 +58,194 @@ static void flush_if_needed(void)
+        print_commands(_builtin_cmds);
     }
 }
 
+/**
+ * Break input line into words, create argv and call the command handler.
+ *
+ * Words are broken up at spaces. A backslash escaped the character that comes
+ * after (meaning it it taken literally and if it is a space it does not break
+ * the word). Spaces can also be protected by quoting with double or single
+ * quotes.
+ *
+ State diagram for the tokenizer:
+```
+
+
+           ┌───[\]────┐   ┌─────["]────┐   ┌───[']─────┐  ┌───[\]────┐
+           ↓          │   ↓            │   │           ↓  │          ↓
+  ┏━━━━━━━━━━┓      ┏━┷━━━━━┓        ┏━┷━━━┷━┓       ┏━━━━┷━━┓     ┏━━━━━━━━━━┓
+  ┃DQUOTE ESC┃      ┃DQUOTE ┠───["]─>┃SPACE  ┃<─[']──┨SQUOTE ┃     ┃SQUOTE ESC┃
+  ┗━━━━━━━━┯━┛      ┗━━━━━━┯┛        ┗┯━━━━┯━┛       ┗━┯━━━━━┛     ┗━━━┯━━━━━━┛
+           │          ↑    │          │    │           │    ↑(store)   │
+           │   (store)│    │   ┌─[\]──┘    └──[*]────┐ │    │          │
+           └────[*]──▶┴◀[*]┘   │                     │ └[*]▶┴◀──────[*]┘
+                               ↓     ┏━━━━━━━┓       ↓
+                               ├◀─[\]┨NOQUOTE┃◀──────┼◀─┐
+                               │     ┗━━━━━┯━┛(store)↑  │
+                               │           │         │  │
+                               │           └─[*]─────┘  │
+                               │     ┏━━━━━━━━━━━┓      │
+                               └────▶┃NOQUOTE ESC┠──[*]─┘
+                                     ┗━━━━━━━━━━━┛
+```
+ */
 static void handle_input_line(const shell_command_t *command_list, char *line)
 {
-    static const char *INCORRECT_QUOTING = "shell: incorrect quoting";
-
     /* first we need to calculate the number of arguments */
-    unsigned argc = 0;
-    char *pos = line;
-    int contains_esc_seq = 0;
-    while (1) {
-        if ((unsigned char) *pos > ' ') {
-            /* found an argument */
-            if (*pos == '"' || *pos == '\'') {
-                /* it's a quoted argument */
-                const char quote_char = *pos;
-                do {
-                    ++pos;
-                    if (!*pos) {
-                        puts(INCORRECT_QUOTING);
-                        return;
-                    }
-                    else if (*pos == '\\') {
-                        /* skip over the next character */
-                        ++contains_esc_seq;
-                        ++pos;
-                        if (!*pos) {
-                            puts(INCORRECT_QUOTING);
-                            return;
-                        }
-                        continue;
-                    }
-                } while (*pos != quote_char);
-                if ((unsigned char) pos[1] > ' ') {
-                    puts(INCORRECT_QUOTING);
-                    return;
+    int argc = 0;
+    char *readpos = line;
+    char *writepos = readpos;
+    enum PARSE_STATE pstate = PARSE_SPACE;
+
+    for (; *readpos != '\0'; readpos++) {
```

Comment:
Initially, this line was `for (bool is_wordbreak; * readpos! = '\ 0'; readpos ++, is_wordbreak = false) {`, but it resulted in a seg fault on native since `is_wordbreak` was not set to false on the next iteration. I think I remember that it worked for nrf52dk and a nucleo board (otherwise I wouldn't have pushed it onto Github, but I don't have a board at home right now to test for it). Are there differences in compiler support regarding the comma operator that I should be aware of? Or did I just misunderstood how the comma operator is working?

Reply:
Honestly no idea, maybe @kaspar030 knows?

Reply:
> Are there differences in compiler support regarding the comma operator that I should be aware of? Or did I just misunderstood how the comma operator is working?

the operator should work the same on all platforms, it is completely standard.

Reply:
> Initially, this line was `for (bool is_wordbreak; * readpos! = '\ 0'; readpos ++, is_wordbreak = false) {`, but it resulted in a seg fault on native since `is_wordbreak` was not set to false on the next iteration.

hm, one issue in that line would be that `is_wordbreak` is undefined in the first iteration, right? could that have caused a segfault?

Reply:
Thank you @kaspar030, that worked out. I just tested it on my esp32 again and it works both ways, but on native only when initializing  `is_wordbreak` explicitly to false.